File: back/go/app/core/database/redis/util/Converter.go
package util

import (
	"errors"
	"go-ttt/app/core/model"
)

func CellPositionToString(position model.CellPosition) (string, error) {
	switch position {
	case model.TL:
		return "TL", nil
	case model.T:
		return "T", nil
	case model.TR:
		return "TR", nil
	case model.L:
		return "L", nil
	case model.C:
		return "C", nil
	case model.R:
		return "R", nil
	case model.BL:
		return "BL", nil
	case model.B:
		return "B", nil
	case model.BR:
		return "BR", nil
	default:
		return "", errors.New("invalid CellPosition")
	}
}

func StringToCellPosition(s string) (model.CellPosition, error) {
	switch s {
	case "TL":
		return model.TL, nil
	case "T":
		return model.T, nil
	case "TR":
		return model.TR, nil
	case "L":
		return model.L, nil
	case "C":
		return model.C, nil
	case "R":
		return model.R, nil
	case "BL":
		return model.BL, nil
	case "B":
		return model.B, nil
	case "BR":
		return model.BR, nil
	default:
		return "", errors.New("invalid CellPosition string")
	}
}

func PlayerTypeToString(playerType model.PlayerType) string {
	return string(playerType.Symbol)
}

func StringToPlayerType(s string) (model.PlayerType, error) {
	switch s {
	case string(model.X):
		return model.PlayerType{Symbol: model.X}, nil
	case string(model.O):
		return model.PlayerType{Symbol: model.O}, nil
	default:
		return model.PlayerType{}, errors.New("invalid PlayerType string")
	}
}

func GameStateToString(gameState model.GameState) string {
	return string(gameState)
}

func StringToGameState(s string) (model.GameState, error) {
	switch s {
	case string(model.InProgress):
		return model.InProgress, nil
	case string(model.Won):
		return model.Won, nil
	case string(model.Draw):
		return model.Draw, nil
	default:
		return "", errors.New("invalid GameState string")
	}
}

File: back/kotlin/app/core/database/redis/util/Converter.kt
package core.database.redis.util

import core.model.CellPosition
import core.model.CellType
import core.model.PlayerType
import core.model.GameState
object Converter {
    
    fun cellPositionToStringRepresentation(cellPosition: CellPosition): String {
        return when (cellPosition) {
            CellPosition.TL -> "TL"
            CellPosition.T -> "T"
            CellPosition.TR -> "TR"
            CellPosition.L -> "L"
            CellPosition.C -> "C"
            CellPosition.R -> "R"
            CellPosition.BL -> "BL"
            CellPosition.B -> "B"
            CellPosition.BR -> "BR"
        }
    }

    fun stringToCellPosition(string: String): CellPosition? {
        return when (string) {
            "TL" -> CellPosition.TL
            "T" -> CellPosition.T
            "TR" -> CellPosition.TR
            "L" -> CellPosition.L
            "C" -> CellPosition.C
            "R" -> CellPosition.R
            "BL" -> CellPosition.BL
            "B" -> CellPosition.B
            "BR" -> CellPosition.BR
            else -> null
        }
    }

    fun stringToPlayerType(string: String): PlayerType? {
        val cellType = when (string) {
            "X" -> CellType.X
            "O" -> CellType.O
            else -> return null
        }
        return PlayerType(cellType)
    }

    fun playerTypeToStringRepresentation(playerType: PlayerType): String {
        return when (playerType.symbol) {
            CellType.X -> "X"
            CellType.O -> "O"
        }
    }

    fun stringToGameState(string: String): GameState {
        return when (string) {
            "IN_PROGRESS" -> GameState.IN_PROGRESS
            "WON" -> GameState.WON
            "DRAW" -> GameState.DRAW
            else -> GameState.IN_PROGRESS
        }
    }

    fun gameStateToStringRepresentation(gameState: GameState): String {
        return when (gameState) {
            GameState.IN_PROGRESS -> "IN_PROGRESS"
            GameState.WON -> "WON"
            GameState.DRAW -> "DRAW"
        }
    }
}